virtualizarr/readers/tiff.py

```diff
@@ -1,79 +1,179 @@
+    if sample_format[0] == 1 and bits_per_sample[0] == 16:
+        return np.dtype(np.uint16)
+    elif bits_per_sample[0] == 64:  # TODO: Check if sample_format matters here
+        return np.dtype(np.float64)
+    else:
+        raise NotImplementedError
+
+
+def _get_codecs(compression):
+    if compression == 8:  # Adobe DEFLATE
+        zlib_props = ZlibProperties(level=6)  # type: ignore
+        conf = dataclasses.asdict(zlib_props)
+        conf["id"] = "zlib"
+    else:
+        raise NotImplementedError
+    codec = registry.get_codec(conf)
+    return codec
 
 
 class TIFFVirtualBackend(VirtualBackend):
     @staticmethod
-    def open_virtual_dataset(
-        filepath: str,
-        group: str | None = None,
-        drop_variables: Iterable[str] | None = None,
-        loadable_variables: Iterable[str] | None = None,
-        decode_times: bool | None = None,
-        indexes: Mapping[str, Index] | None = None,
-        virtual_backend_kwargs: Optional[dict] = None,
-        reader_options: Optional[dict] = None,
-    ) -> Dataset:
-        if virtual_backend_kwargs:
-            raise NotImplementedError(
-                "TIFF reader does not understand any virtual_backend_kwargs"
-            )
+    def _construct_chunk_manifest(
+        ifd: ImageFileDirectory,
+        *,
+        path: str,
+        shape: tuple[int, ...],
+        chunks: tuple[int, ...],
+    ) -> ChunkManifest:
+        tile_shape = tuple(math.ceil(a / b) for a, b in zip(shape, chunks))
+        # See https://web.archive.org/web/20240329145228/https://www.awaresystems.be/imaging/tiff/tifftags/tileoffsets.html for ordering of offsets.
+        tile_offsets = np.array(ifd.tile_offsets, dtype=np.uint64).reshape(tile_shape)
+        tile_counts = np.array(ifd.tile_byte_counts, dtype=np.uint64).reshape(
+            tile_shape
+        )
+        paths = np.full_like(tile_offsets, path, dtype=np.dtypes.StringDType)
+        return ChunkManifest.from_arrays(
+            paths=paths,
+            offsets=tile_offsets,
+            lengths=tile_counts,
+        )
 
-        from kerchunk.tiff import tiff_to_zarr
+    @staticmethod
+    async def _open_tiff(*, path: str, store: SupportedStore) -> TIFF:
+        from async_tiff import TIFF
 
-        if reader_options is None:
-            reader_options = {}
+        return await TIFF.open(path, store=store)
 
-        reader_options.pop("storage_options", {})
-        warnings.warn(
-            "storage_options have been dropped from reader_options as they are not supported by kerchunk.tiff.tiff_to_zarr",
-            UserWarning,
+    @staticmethod
+    def _construct_manifest_array(
+        *, ifd: ImageFileDirectory, path: str
+    ) -> ManifestArray:
+        if not ifd.tile_height or not ifd.tile_width:
+            raise NotImplementedError(
+                f"TIFF reader currently only supports tiled TIFFs, but {path} has no internal tiling."
+            )
```

Comment:
Possibly ignorant question, but can't we represent a non-tiled TIFF as a single virtual chunk?

Reply:
there are three categories:
- no chunks (yes, would map to a single virtual chunk)
-  tiled TIFFs (as implemented)
- striped TIFFs (basically only chunked along the `y` dimension

None of these are hard to represent as a manifest array, but it's just a matter of finding/using the proper tags to determine the chunk structure

Reply:
Do the striped TIFFs have compression applied independently along each line?

Reply:
yes, striped as single lines or in a group, same as any other tiling